src/esp/sim/Simulator.cpp

```diff
@@ -389,5 +414,134 @@ bool Simulator::recomputeNavMesh(nav::PathFinder& pathfinder,
   return true;
 }
 
+// Agents
+void Simulator::sampleRandomAgentState(agent::AgentState::ptr agentState) {
+  agentState->position = pathfinder_->getRandomNavigablePoint();
+  const float randomAngleRad = random_.uniform_float_01() * M_PI;
+  quatf rotation(Eigen::AngleAxisf(randomAngleRad, vec3f::UnitY()));
+  agentState->rotation = rotation.coeffs();
+  // TODO: any other AgentState members should be randomized?
+}
+
+agent::Agent::ptr Simulator::addAgent(
+    const agent::AgentConfiguration& agentConfig,
+    scene::SceneNode& agentParentNode) {
+  // initialize the agent, as well as all the sensors on it.
+
+  // attach each agent, each sensor to a scene node, set the local
+  // transformation of the sensor w.r.t. the agent (done internally in the
+  // constructor of Agent)
+
+  auto& agentNode = agentParentNode.createChild();
+  agent::Agent::ptr ag = agent::Agent::create(agentNode, agentConfig);
+
+  // Add a RenderTarget to each of the agent's sensors
+  for (auto& it : ag->getSensorSuite().getSensors()) {
+    if (it.second->isVisualSensor()) {
+      auto sensor = static_cast<sensor::VisualSensor*>(it.second.get());
+      renderer_->bindRenderTarget(*sensor);
+    }
+  }
+
+  agents_.push_back(ag);
+  // TODO: just do this once
+  if (pathfinder_->isLoaded()) {
+    ag->getControls()->setMoveFilterFunction(
+        [&](const vec3f& start, const vec3f& end) {
+          return pathfinder_->tryStep(start, end);
```

Comment:
This won't be synchronized with python move filter functions.

Reply:
Python code doesn't use this pathfinder. Cool if we address this later when python is modified to use this pathfinder?

Reply:
Sounds good.